xdsl/transforms/jax_use_donated_arguments.py

```diff
@@ -0,0 +1,144 @@
+import itertools
+from collections.abc import Sequence
+from dataclasses import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import builtin
+from xdsl.dialects.bufferization import MaterializeInDestinationOp
+from xdsl.dialects.builtin import FunctionType, TensorType
+from xdsl.dialects.func import FuncOp, ReturnOp
+from xdsl.ir import BlockArgument, Operation, SSAValue
+from xdsl.irdl import VarOperand
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+def map_outputs_to_inputs(
+    donated_inputs: Sequence[BlockArgument], outputs: VarOperand
+) -> dict[int, int]:
```

Comment:
It's not super clear by reading the function name what maps to what. I'm tepted to try mapping the ssa values directly so the type communicates in which direction the mapping goes, maybe something like this?
```suggestion
def donated_input_by_output(
    donatable_inputs: Sequence[BlockArgument], outputs: Sequence[SSAValue]
) -> dict[SSAValue, BlockArgument]:
```